### PR TITLE
Add CodeRequiredHook

### PR DESCRIPTION
**Pull request by NickCuso**
# Description

Require the user know a code in order to purchase a key.

The test shows the use case where the lock is free if you know the code.  I'm thinking we can also offer a second lock, which has a non-zero keyPrice that people can purchase if they don't have the code.

e.g. blog about an honeypot example:

HoneyPot blog
A few paragraphs about the scam, links to a specific example, and some clues about what's going on.

Then behind the lock is a few paragraphs about with proof of manipulation and various ways to detect the scam yourself.

You can buy a key for $1 or it's free if you can find the tx that turned our contract that looks like a simple game into a honeypot scam.  Enter tx hash [0x..........]

--

Idea from #ideas slack channel

# Issues

Fixes #
Refs #

# Checklist:

- [x] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [x] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [x] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

**Comment by NickCuso** [comment]
This contract works for a more standard use case as well: a simple redeem code.  One lock that's free (or discounted) if you have the code, and another that's paid if you don't.

**Comment by NickCuso** [comment]
BTW we can also create a similar contract which offers discount codes.  e.g. enter `unlock-10` for 10% off.  We implement that by withdrawing funds from the lock and transferring them to the buyer.  So it's not the best solution - you pay full price and then get an instant refund.

**Comment by julien51** [comment]
Very cool! @nfurfaro I'll trust your judgement on the solidity parts!

**Comment by nfurfaro** [comment]
>Require the user know a code in order to purchase a key.

What sor t of "code" are we thinking here? a tx hash as per the Ideas slack channel, or something different?

File: smart-contract-extensions/contracts/unlock-examples/hooks/CodeRequiredHook.sol [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+import '@openzeppelin/contracts/cryptography/ECDSA.sol';
+import '@openzeppelin/contracts/access/roles/WhitelistAdminRole.sol';
+import '@openzeppelin/contracts/introspection/IERC1820Registry.sol';
+import 'hardlydifficult-ethereum-contracts/contracts/proxies/CallContract.sol';
+
+/**
+ * @notice Used with a Lock in order to require the user knows
+ * a code in order to buy.
+ * @dev Set this contract as the beneficiary of the lock.
```
Comment (nfurfaro) [inline smart-contract-extensions/contracts/unlock-examples/hooks/CodeRequiredHook.sol]: So here we set the CodeRequiredHook contract as the beneficiary, but I'm not seeing how changing the beneficiary(as per comments for proxyCall) would remove the code requirement? I would have thought we'd need to update the call to this hook from the purchase function.

**Comment by NickCuso** [comment]
The code can be anything - that's determined by the frontend implementation.  The test shows a string and that string could be any length - it could represent a redeem code or an answer to a question.  The frontend could lower().trim() ect to ensure things like casing do not cause it to fail.

The code could also be a number or whatever.

By feeding it through a hash, any type with effectively unlimited inputs is fine.

File: smart-contract-extensions/contracts/unlock-examples/hooks/CodeRequiredHook.sol (hunk 1 above)
Comment (NickCuso) [reply smart-contract-extensions/contracts/unlock-examples/hooks/CodeRequiredHook.sol]: We check for a hook today by asking erc-1820 if the beneficiary has registered one.

Comment (NickCuso) [reply smart-contract-extensions/contracts/unlock-examples/hooks/CodeRequiredHook.sol]: We could either make this contract the beneficiary, or instead have the beneficiary call erc1820 to say this is the hook to use.

The downside with the latter is the beneficiary's other locks will call the hook as well.  So the hook needs to check if it's the lock of interest... which is what motivated our desire to remove 1820 completely.

Comment (nfurfaro) [reply smart-contract-extensions/contracts/unlock-examples/hooks/CodeRequiredHook.sol]: Ah, right. So while the new hook contract is the beneficiary, there is a registered hook (itself). Changing the beneficary means the new beneficiary won't have (by default) a registered hook, and the purchase will proceed as a normal purchase.

One question about this: If all purchase calls have to check the erc-1820 contract for registered hooks, what sort of gas-penalty do users pay for this in the cases where a hook is not used?

Comment (NickCuso) [reply smart-contract-extensions/contracts/unlock-examples/hooks/CodeRequiredHook.sol]: That's right.

It's a non-zero cost to use 1820 for sure, another motivation to change.

File: smart-contract-extensions/test/unlock-examples/hooks/codeRequiredHook.js [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  const keyBuyer = accounts[5]
+  let lock
+  let hookContract
+  // Never expose the answer or its PK directly in JS for any integrations
+  // Only the answer account's address can safely be exposed
+  let answerAccount
+
+  beforeEach(async () => {
+    // Ganache only setup (a no-op if erc1820 already exists)
+    await erc1820.deploy(web3)
```
Comment (nfurfaro) [inline smart-contract-extensions/test/unlock-examples/hooks/codeRequiredHook.js]: Maybe use the oz-testhelper here? 
https://docs.openzeppelin.com/test-helpers/0.5/api#erc1820

Comment (NickCuso) [reply smart-contract-extensions/test/unlock-examples/hooks/codeRequiredHook.js]: Cool - didn't know that was a thing.

But it's a different use case.  We only use interfaceIds in Solidity for this example... I never use them in JS here.

The line you tagged is about deploying the erc1820 contract to ganache

File: smart-contract-extensions/test/unlock-examples/hooks/codeRequiredHook.js [hunk 3]
```diff
@@ -0,0 +1,110 @@
+const { constants, protocols } = require('hardlydifficult-ethereum-contracts')
+const erc1820 = require('erc1820')
+const { reverts } = require('truffle-assertions')
+
+const CodeRequiredHook = artifacts.require('CodeRequiredHook.sol')
+
+contract('CodeRequiredHook', accounts => {
+  const lockOwner = accounts[1]
```
Comment (nfurfaro) [inline smart-contract-extensions/test/unlock-examples/hooks/codeRequiredHook.js]: Minor, but could we maybe use lockCreator here moving forward? 
We'll be doing a fairly large migration away from using "owner" fairly soon. 
"lockCreato"r is pretty descriptive, and may help with the transition away from using Ownable, at least cognitively.

**Comment by NickCuso** [comment]
Yes! This contract supports discount codes already - if we offer one lock with a code required that's discounted and another lock that's full price.

File: smart-contract-extensions/test/unlock-examples/hooks/codeRequiredHook.js (hunk 2 above)
Comment (nfurfaro) [reply smart-contract-extensions/test/unlock-examples/hooks/codeRequiredHook.js]: oops, sorry. meant to link to this:
https://docs.openzeppelin.com/test-helpers/0.5/api#erc1820registry

Comment (NickCuso) [reply smart-contract-extensions/test/unlock-examples/hooks/codeRequiredHook.js]: OZ is great and all but I'm going to leave this as is for the moment.  It's the same feature, but the library I'm using is the official example from the EIP.

If/when we make this change we can do smart-contracts and smart-contract-extensions at the same time.

Comment (NickCuso) [reply smart-contract-extensions/test/unlock-examples/hooks/codeRequiredHook.js]: Oh! I just remembered this step is optional cause `createTestLock` will take care of it.  I'll delete to simplify the test.

**Pull request merged**